proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+Fortified Language Editions
+===========================
+
+.. author:: Richard Eisenberg (with some contributions by Simon Peyton Jones)
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/636>`_.
+.. sectnum::
+.. contents::
+
+.. _manifesto: https://github.com/ghc-proposals/ghc-proposals/pull/628
+
+I propose to lean more heavily on *language editions*. This proposal describes
+how a language edition can control more than just extensions (importantly, one
+can control *warnings* as well), and proposes several concrete language editions.
+One of the proposals is for ``Stable2024``, which will continue to be supported,
+without change to the set of user programs accepted, for at least 3 years.
+
+This is extracted from a longer `manifesto`_ about how GHC should treat language
+extensions. For some introductory text describing the mindset that leads to this
+design, readers may want to skim the introductory sections of that manifesto.
+However, the effective payload of this proposal is all in this one document.
+
+Motivation
+----------
+
+The primary motivation behind the use of language editions is that they
+can succinctly inform GHC what kind of user it's faced with, so GHC
```

Comment:
I think the focus on the user is good, but what @tomjaguarpaw 's comment points out is that we haven't worked out what the (sorry) _user personas_ are. What does the user of `Stable2024` _want_ or expect? In particular, I think there are a range of possible expectations:
- About their own code:
    - Is guaranteed to be stable on later versions of GHC
    - Is likely to be stable on later versions of GHC
    - Can/cannot have those guarantees reduced if they make exceptions
- About their dependencies:
    - No assumptions
    - All/some are guaranteed to be stable on later versions of GHC
    - All/some are likely to be stable on later versions of GHC

Here are some possible combined positions:

1. "I want tools to help me ensure that the code that I write will be stable on later GHCs, but I want the ability to take things into my own hands and make exceptions if that is useful. I want to be able to communicate stability levels between packages, so that I can have a level of assurance that my dependencies are stable, and communicate my own guarantees downstream; but I am okay with this being not a guarantee (since I expect everyone to make exceptions here and there)"
2. "I want everything to be rock solid. I have no interest in using cutting-edge features at all, and will jump through as many hoops as necessary to avoid it. I also really don't want to deal with dependency issues, so I only want to use dependencies that are rock solid as well"
3. "I mostly don't care that much, but it would be nice if the defaults pushed packages towards using more stable features so things were less likely to break"

Similarly for students:

1. "I have no idea what I'm doing. I just want it to work so I can clear this class. I'm not going to do anything the teacher doesn't tell me to and I want no surprises."
2. "I love Haskell! This is great! I immediately want to play with all the toys! It's day 2 and I'm trying to implement an iceland_jack tweet!"

I'm unsure who the users of `Experimental` or `Latest` would be. I guess people who don't care about stability at all. But most people care a bit. In particular, here is one more persona I think we might want to worry about:

1. "I really need to use `LinearTypes`. That's kind of the point of this project. I know that opens me up to a bit of breakage, but that doesn't mean I want _everything_ to break. If anything, I want to keep the rest as stable as possible to compensate for my big stability hit in using `LinearTypes`!"

----

The reason that I think this is important is that while I appreciate the desire to be opinionated, I think there's a risk that we fixate on a particular manifestation of the "cares about stability" persona or the "is a student" persona, and that can also be pretty harmful to the other people!

Reply:
I find this really helpful.

For the different stability goals: I continue to claim that `Experimental2024` is a viable "escape hatch". Maybe we just need `TryingToBeStable2024` which is technically identical to `Experimental2024` but socially distinct? (That's a bit of a joke.) That is, in order to suit the "rock solid" camp, we need `Stable2024` to be completely reliable (or as close as we can make it). The other stability folks can work with `Experimental2024` and social cues. I do like the `LinearTypes` example. But that author still has to make a choice: do they want rock-solid dependencies (`Stable2024`) or not (`Experimental2024`)? Anything beyond rock-solid requires the social cues again.

I continue to become less enthusiastic about `Student`. I agree that the error-message aspect could be controlled using a different, less heavy-handed mechanism.

For `Latest`, the persona is simple: "I just want to experiment with Haskell. I'm not going to publish this." Except that maybe they do, in the end, for use by other `Latest` people.

Reply:
This is an excellent point. This proposal offers no enforcement mechanism around dependencies. I think it probably should. Specifically, I think there should be an option `stable`, specified in a cabal file, that requires all dependencies to also be `stable`. It would be an error to specify `stable` and not have a `StableXXX` language edition selected. For simplicity, I'm suggesting here that the cabal flag not specify the year; that can be done in the language edition.

* A cabal file (or individual module) can specify `Stable2024` *without* specifying `stable`. That's fine. It means that the package or module is *locally stable* but not part of the guaranteed-stable ecosystem.
* This plan is *much simpler* than the plan around Safe Haskell, because there is no escape hatch. (Cue the tomatoes, which I deflect below.) That is, there is no equivalent to `Trustworthy`. Simple straightforward transitive dependencies. No exceptions.
* (Here is where I deflect tomatoes.) The reason this works is that we can all program without the use of unstable extensions. Look at the "N"s in the `Stable` column in the proposal. The key question: can we live without these extensions? In particular, can we compile major packages (including `base`!) without these? Let's look.

These N extensions affect only concrete syntax; other concrete syntax is easily available: `AlternativeLayoutRule`, `AlternativeLayoutRuleTransitional`,`CUSKs`, `DeriveAnyClass`, `NondecreasingIndentation`, `NPlusKPatterns`, `NumDecimals`, `OverlappingInstances`, `OverloadedRecordUpdate`, `ParallelArrays`, `RelaxedLayout`, `TransformListComp`, `TypeAbstractions`

These N extensions do nothing or are subsumed by other extensions: `AutoDeriveTypeable`, `NullaryTypeClasses`, `TypeInType`

These N extensions actually do increase expressiveness, but in ways we actively do not want: `DatatypeContexts`

These N extensions actually do increase expressiveness, but we have better ways to say it now: `DefaultSignatures`

These N extensions are properly experimental: `LinearTypes`, `RequiredTypeArguments`, `StaticPointers`

These N extensions are part of Safe Haskell: `Safe`, `Trustworthy`, `Unsafe`

Safe Haskell aside, the question really is: do we want stable packages to depend on other packages using `DatatypeContexts`, `DefaultSignatures`, `LinearTypes`, `RequiredTypeArguments`, or `StaticPointers`? Those are really the only extensions in question. `DatatypeContexts` have been deprecated for more than a decade. That said, they're not unstable; I could see allowing these in `Stable2024` if we want to extend the deprecation period. (Though, actually, I think `DatatypeContexts` can be perfectly simulated by pattern synonyms. So maybe we could indeed excise them from the language.) `DefaultSignatures` are more controversial. We should probably do a little impact study to see how hard it would be to get rid of them (in favor of the more flexible `DerivingVia`). I'm content to allow these into `Stable2024`. The other three really are experimental, and I don't think we should yet allow them into a `stable` ecosystem. (`TypeAbstractions` is both experimental and has syntactic workarounds.)

Safe Haskell is a different ball of wax. Maybe the best approach is to deprecate them, but still allow them in `Stable2024`, albeit that they would have no effect and cause warnings.

The bottom line for me here is that I think we can make this `stable` ecosystem without need a trust mechanism. It's the trust mechanism that makes Safe Haskell hard. It's necessary there. But I think it's not necessary here.

> So who wants `Latest`?

I do! :) That is, I'd specify `Latest` in my little Haskell experiments. But I easily admit that I don't need to.

Reply:
> A cabal file (or individual module) can specify Stable2024 without specifying stable

That's great. That's the way I'll use it (and the way I expect the vast majority of users would want to use it)

> I think there should be an option stable, specified in a cabal file, that requires all dependencies to also be stable. It would be an error to specify stable and not have a StableXXX language edition selected

To what extent is this a GHC concern? The cabal team are welcome to implement that, of course, but that's quite independent of what the GHC steering committee decides (unless you're suggesting some sort of joint proposal).

Reply:
> This proposal offers no enforcement mechanism around dependencies. I think it probably should. Specifically, I think there should be an option `stable`, specified in a cabal file, that requires all dependencies to also be `stable`.

What problem is this intended to solve, and for which class of users? How often are users accidentally depending on libraries that break because they used one of these extensions?

My impression is that the main reason dependencies lead to stability problems is not principally due to language changes, but library changes (and https://github.com/ghc-proposals/ghc-proposals/pull/636#issuecomment-1983313974 seems to back this impression up). So having a `stable` option that increases the chances of language stability throughout the dependency stack, but doesn't do anything about libraries changing their APIs, may not have the impact we'd like.